core/src/main/scala/dagr/core/exec/TaskCache.scala

```diff
@@ -0,0 +1,323 @@
+
+  /** The set of tasks that should be executed. */
+  private val tasksToExecute: mutable.Set[Task] = new mutable.HashSet[Task]()
+
+  /** The set of tasks whose sub-tree should always be executed, since the task or an ancestor of the task could not be
+    * mapped to a definition in the replay log. */
+  private val missingAncestors: mutable.Set[Task] = new mutable.HashSet[Task]()
+
+  /** Returns true if the task should be executed, false otherwise. */
+  def execute(task: Task): Boolean = this.tasksToExecute.contains(task)
+
+  /** Returns the tasks that have been registered. */
+  protected def taskSet: Set[Task] = this.taskToDefinition.keySet.toSet
+
+  /** Returns true if the task has been registered, false otherwise. */
+  protected def registered(task: Task): Boolean = this.taskToDefinition.contains(task)
+
+  /** Registers the root task, namely a task that has no tasks (and never had any tasks) that depend on it. */
+  protected def registerRoot(root: Task): Unit = {
+    // find the status in the replay log and decide what to do.
+    val rootDefinition = Definition.buildRootDefinition(root)
+    this.taskToDefinition(root) = rootDefinition
+    // NB: the root always has the same parent code!
+    // NB: assumes only one root
+    this.definitions.find { d => d.parentCode == rootDefinition.parentCode } match {
+      case None                   => unreachable(s"Could not find a root task in $replayLog")
+      case Some(replayDefinition) => // make a decision
+        this.definitionMapping(rootDefinition) = replayDefinition
+        // NB: don't determine if it should be executed yet; register should be called again either with the root's
+        // children, or root itself as its own child (a unit task).
+    }
+  }
+
+  /** Registers the given children of the parent, assuming the children are the task built by the parent (and does not
+    * include the parent itself) */
+  protected def registerChildren(parent: Task, currentChildren: Seq[Task]): Unit = {
+    val parentDefinition = currentDefinition(parent)
+
+    // Create definitions for reach child
+    val currentChildDefinitions = currentChildren.zipWithIndex.map { case (child, childNumber) =>
+      buildDefinition(child, parentDefinition.code, childNumber)
+    }.toIndexedSeq
+
+    // Always execute the children if the parent has a missing ancestor (or is the missing ancestor itself).
+    if (hasMissingAncestor(parent)) {
+      setMissingAncestorAndExecute(currentChildren)
+    }
+    else {
+      // Get the children of the parent from the previous execution
+      val previousChildDefinitions = this.previousChildDefinitions(parent)
+
+      // Check to see if any of the current child definitions are equivalent
+      val equivalentChildren = currentChildDefinitions.exists { left =>
+        currentChildDefinitions.exists { right => left != right && left.equivalent(right) }
+      }
+
```

Comment:
@tfenne this is the where we determine if a pipeline-like task (a task that returns other tasks) is equivalent across executions.  We look for:
1. The task must produce the same # of children.
2. None of the children produced are "equivalent" to each other (i.e. one cannot be misinterpreted for the other across executions).
3. Each child has one and only one equivalent "definition" from the previous execution, meaning for each child task we found one and only one potential matching task from the previous execution.

If the above are true, we evaluate whether or not to run the children separately (by their previous status), otherwise, we force all the children to run.